File: solution/1382. Balance a Binary Search Tree/1382.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* balanceBST(TreeNode* root) {
        vector<TreeNode*> A;
        dfs(root, A);
        return build(A, 0, A.size() - 1);
    }

private:
    void dfs(TreeNode* node, vector<TreeNode*>& A) {
        if (!node) return;
        dfs(node->left, A);
        A.push_back(node);
        dfs(node->right, A);
    }
    
    TreeNode* build(vector<TreeNode*>& A, int l, int r) {
        if (l > r) return nullptr;
        
        int mid = l + (r - l) / 2;
        auto node = A[mid];
        node->left = build(A, l, mid - 1);
        node->right = build(A, mid + 1, r);
        return node;
    }
};

File: solution/1382. Balance a Binary Search Tree/1382.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def balanceBST(self, root: TreeNode) -> TreeNode:
        A = []
        
        def dfs(node):
            if not node: return None
            dfs(node.left)
            A.append(node)
            dfs(node.right)
        
        dfs(root)
        
        def build(A, l, r):
            if l > r: return None
            
            mid = (l + r) // 2
            A[mid].left = build(A, l, mid - 1)
            A[mid].right = build(A, mid + 1, r)
            return A[mid]
        
        return build(A, 0, len(A) - 1)
    
                
